src/box/xlog.h

```diff
@@ -215,7 +241,16 @@ static inline bool
 xdir_has_garbage(struct xdir *dir, int64_t signature)
 {
 	struct vclock *vclock = vclockset_first(&dir->index);
-	return vclock != NULL && vclock_sum(vclock) < signature;
+	if (vclock == NULL)
+		return false;
+
+	struct vclock retention_vclock;
+	xdir_get_retention_vclock(dir, &retention_vclock);
+	if (vclock_is_set(&retention_vclock))
+		/** retention_vclock is >= vclock. */
+		return vclock_sum(&retention_vclock) < signature;
```

Comment:
Wait. When `xdir_has_garbage` returns true, it means that we may delete something.

But `vclock_sum(&retention_vclock) < signature` means that `signature` is protected by retention vclock.

Shouldn't it be
```c
if (vclock_is_set(&retention_vclock)) {
    return vclock_sum(&retention_vclock) >= signature && vclock_sum(vclock) < signature;
}
return vclock_sum(vclock) < signature;
```
?

Reply:
You're right, it won't work. `xdir_has_garbage` always returns false, when retention is used. However, your approach is not fully correct too. Even though it will work now, when `xdir_has_garbage` is used only in one place, inside `wal_fallocate` (`signature` is always retention), it won't in general. 

---

When there's `retention_vclock` set, we may have the following situations:

1. Situation 1. `vclock_sum(vclock)` is 10, `vclock_sum(retention_vclock)` is 20. `signature` is 21.
```
index                     10    20    30    ...
retention                       20    30    ...
```

Your return value will be `false`, as retention sum is < than signature, but we have garbage, which is 10. My return value is correct.

2. Situation 2. The same, as in the 1st one, but `signature` is 19. Your return value correctly returns true, mine - false.
3. Sitation 3.  `vclock_sum(vclock)` is 10, `vclock_sum(retention_vclock)` is 10. `signature` is 11.
```
index                     10    20    30    ...
retention                 10    20    30    ...
```
Mine returns true, which is wrong, no garbage. Your - false, which is correct. 

----

I suppose, we must introduce more general solution, which works in all cases. If retention is used, we must find a vclock, which is less, than a signature and which is not protected by gc. If retention is set and it's not the first in the index, then it's enough to check, that the first index is less than a signature. If retention is used, but retention == to the first index, then all files are protected. Please, check implementation.

I also added unit tests (in EE part), as I made such a critical error in the function. Thank you for noticing!

Reply:
Thanks for paying attention! I also made a mistake.

The current version looks good to me, but I suggest applying this small diff. It doesn't change anything now, because `retention` can't be smaller or incomparable with `vclock`, but still looks safer.
```diff
diff --git a/src/box/xlog.h b/src/box/xlog.h
index 5310ac52c0..b60e30d335 100644
--- a/src/box/xlog.h
+++ b/src/box/xlog.h
@@ -246,8 +246,10 @@ xdir_has_garbage(struct xdir *dir, int64_t signature)
 
        struct vclock retention;
        xdir_get_retention_vclock(dir, &retention);
-       if (!vclock_is_set(&retention) || vclock_compare(&retention, vclock) != 0)
+       if (!vclock_is_set(&retention) ||
+           vclock_compare(vclock, &retention) < 0) {
                return vclock_sum(vclock) < signature;
+       }
 
        /** All files are protected from gc. */
        return false;
```

Reply:
Sure, agree. Fixed